Edge scrolling

logiclrd opened the pull request:
This PR adds a new scroll style: Edge scrolling.

https://github.com/user-attachments/assets/eac377e9-1848-48e8-b596-9a6a29bd315d

Edge scrolling mode behaves like Scrollbar mode, with the added function that when the mouse cursor approaches the edge of the viewport, the view automatically scrolls in that direction if possible. This allows for convenient scrolling around the view while maintaining stability in the view position, as compared to the Auto scrolling mode in which every mouse cursor position in the view corresponds to a unique scroll offset.

Closes: #3053

## Implementation details:

### "Bump Mouse"

For edge scrolling to have a natural "feel" to it, the mouse cursor must "bump" against the edge. That is, when scrolling occurs as a result of edge scrolling, the mouse cursor should stay anchored to the point in the underlying view that it is overtop of before the scroll. In order to achieve this, it is necessary to force the position of the mouse cursor to change, called "warping" the mouse cursor in some toolkits.

This functionality is not a part of the base feature set exposed by Flutter, so it must be implemented by means of a Platform Channel. I noticed that the code already has a Platform Channel, but only for the MacOS platform. This PR co-opts that platform, so that there aren't _two_ platform channels when running on MacOS. The name of the channel within its namespace is changed from `macos` to simply `host`.

I did my best to implement the functionality for all three platform layers. I am currently developing on a Linux host, and was able to test and verify that the Gdk-based implementation of `bumpMouse` works, at least on my system running KDE. I have access to a Windows host and should be able to verify that implementation later, but it is as yet untested.

_I do not have access to an OS X development environment and have no way to test the `macos` implementation._ I wouldn't be surprised if it doesn't even build. I have no clue what I'm doing with Swift code :slightly_smiling_face: 

The Platform Channel can only be accessed from the root isolate, so this PR also adds a message type to the existing window event channel system so that the isolates hosting remote desktop windows can direct their `bumpMouse` actions properly.

### `ScrollStyle`

I noticed a number of places independently implementing conversion of `ScrollStyle` values, in one place to/from JSON and in another to/from an arbitrary string form for method channel encoding. Before proceeding, I centralized the formatting logic.

### Configuration

I have at this point not investigated the possibility of making the width of the edge that triggers scrolling configurable. It can't be too small, because if the mouse is moved quickly enough, it can skip it entirely between events and escape the window. I'm not convinced that this is a bug. In my debugging, the value of 120 that is presently hardcoded worked pretty well, for me on my trackpad at least.

### Automated Testing

There are some things conducive to automated testing, and I haven't written any of those yet at this point. With a bit of abstraction, the edge scrolling algorithm itself wouldn't be a bad thing to have tests for.

logiclrd committed:
Repurposed the MacOS-specific platform channel mechanism for all platforms:
- Renamed the channel from "org.rustdesk.rustdesk/macos" to "org.rustdesk.rustdesk/host".
- Renamed _osxMethodChannel in platform_channel.dart to _hostMethodChannel.
- Updated linux/my_application.cc to use the fl_* API to set up a Method Channel and to dispose it during my_application_dispose.
- Updated windows/runner/flutter_window.cpp to use the C++ API to set up a Method Channel.
- Updated the channel name in macos/Runner/MainFlutterWindow.swift.

Signed-off-by: Jonathan Gilbert <logic@deltaq.org>

logiclrd committed:
Added a method "bumpMouse" to the Platform Channel.
Added a thunk to call the method through the channel to platform_channel.dart.
Added implementation bump_mouse() in linux/my_application.cc using Gdk API calls. Updated host_channel_call_handler to process "bumpMouse" method call messages by calling bump_mouse.
Added implementation Win32Desktop::BumpMouse in windows/runner/win32_desktop.cpp/.h.  Updated the inline method call handler in flutter_window.cpp to handle "bumpMouse" method calls by calling Win32Desktop::BumpMouse.
Updated the method call handler in macos/Runner/MainFlutterWindow.swift to handle "bumpMouse" method call messages. Updated MainFlutterWindow to use a subclass of FlutterViewController exposing access to mouseLocationOutsideOfEventStream.

Signed-off-by: Jonathan Gilbert <logic@deltaq.org>

logiclrd committed:
Added message type kWindowBumpMouse to the multiwindow window event model:
- Added constant kWindowBumpMouse to consts.dart.
- Updated the method handler attached to rustDeskWinManager by DesktopHomePageState to recognize kWindowBumpMouse and translate it to a call to RdPlatformChannel.bumpMouse.

Signed-off-by: Jonathan Gilbert <logic@deltaq.org>

logiclrd committed:
Centralized serialization of ScrollStyle values, moving JSON and string conversions into methods toString/fromString and toJson/fromJson within the type.

Signed-off-by: Jonathan Gilbert <logic@deltaq.org>

logiclrd committed:
Added new scroll style for edge scrolling:
- Added ScrollStyle enum member "scrolledge". Added corresponding constant kRemoteScrollStyleEdge to consts.dart for the string serialized form.
- Updated sites checking specifically for ScrollStyle.scrollbar to instead check for NOT ScrollStyle.scrollauto.
- Added radio buttons for the new "ScrollEdge" style to desktop_setting_page.dart and remote_toolbar.dart. Added new string "ScrollEdge" to lang/template.rs.

Signed-off-by: Jonathan Gilbert <logic@deltaq.org>

logiclrd committed:
Implemented edge scrolling:
- Added methods edgeScrollMouse and pushScrollPositionToUI to class CanvasModel in model.dart.
- Added boolean parameter edgeScroll to handleMouse, handlePointerDevicePos and processEventToPeer in input_model.dart.
- Updated handlePointerDevicePos in input_model.dart to call edgeScrollMouse on move events when the edgeScroll parameter is true.
- Added convenience accessor useEdgeScroll to the InputModel class. Updated call sites to handleMouse to use it to supply the value for the edgeScroll parameter.

Signed-off-by: Jonathan Gilbert <logic@deltaq.org>

logiclrd commented:
Outstanding issue: If the cursor strays into the client area while the window frame is being dragged in a resize operation, then the mouse movement gets processed as an edge scroll action and `bumpMouse` moves the cursor, which causes a feedback effect of window resize effects that rapidly squishes the window down to minimum size. To prevent this, I need edge scrolling, or at the very least `bumpMouse` actions, to be suppressed during resize drag operations. I'm not sure if Flutter itself exposes this information.

logiclrd committed:
Updated CanvasModel.edgeScrollMouse to be resilient to receiving events when _horizontal/_vertical aren't wired up to any UI.

logiclrd committed:
Updated CanvasModel to take notifications of resizes via method notifyResize and to suppress edge scrolling briefly after a resize.
Updated the onWindowResized handler in tabbar_widget.dart to call notifyResize on the canvasModel of any RemotePage tabs.

logiclrd commented:
Okay that fixes the resize glitch. :slightly_smiling_face:

logiclrd committed:
Half a go at fixing MainFlutterWindow.swift.

logiclrd committed:
Copilot feedback.

rustdesk commented:
ChatGPT? :(

logiclrd committed:
Applied fix suggested by Copilot in its explanation of the build error.

logiclrd committed:
Fixed a couple of silly errors in windows/runner/flutter_window.cpp.

logiclrd committed:
Fixed MainFlutterWindow.swift build errors.

Co-Authored-By: fufesou <linlong1266@gmail.com>
Signed-off-by: Jonathan Gilbert <logic@deltaq.org>

logiclrd committed:
Moved new translation to the end of template.rs.
Reran res/lang.py.

Signed-off-by: Jonathan Gilbert <logic@deltaq.org>

logiclrd commented:
> This function on macOS does not work.

It looks like the snippet I found for retrieving the current mouse position isn't working.

I found a discussion elsewhere on the Internet that said that you can just poke into `self.view.window.mouseLocationOutsideOfEventStream` in the context of the view controller. I copy/pasted their code.

The core of the issue comes down to the fact that the mouse position is coming from one place and the functionality for warping the mouse to a new position is in a different place, and these subsystems aren't working with the same coordinate space. But, I don't know the system, and I don't have the means to experiment.

In short: The code expects that `flutterViewController.mouseLocation` will return the absolute position of the cursor within the coordinate system that encapsulates all screens. It then has to contend, from the things I read, with the fact that Core Graphics for some reason uses an inverted coordinate system with the Y axis pointing upward and the origin in the lower-left of the primary monitor, so it converts the coordinate into that space before applying the delta. Maybe that conversion isn't actually needed??

The function `CGDisplayMoveCursorToPoint` is clearly documented to take a point in the absolute coordinate space of the displays, anchored at the primary screen's origin. So, logically, it must be that the code that establishes the current cursor position based on `mouseLocationOutsideOfEventStream` isn't getting the expected value.

<hr />

And I've just read the documentation for `mouseLocationOutsideOfEventStream` and it is explicitly documented as being relative to the receiving window's coordinates. There is a static property on `NSEvent` called `mouseLocation` documented as returning the position in screen coordinates. So, let's give that a go. :slightly_smiling_face: 

Google's AI result suggests also that the Y coordinate conversion isn't needed. I'll push up a commit that uses `NSEvent.mouseLocation` with no conversion. Crossing fingers. :crossed_fingers:

logiclrd committed:
Switched MainFlutterWindow.swift to use NSEvent.mouseLocation.

logiclrd committed:
Updated MainFlutterWindow.swift code based on build error.

logiclrd committed:
Fixed silly typo.

logiclrd commented:
@fufesou Please give the latest code a try when it's convenient and let me know if it behaves better on OS X. 🙂

fufesou commented:
https://github.com/user-attachments/assets/4b702fc7-41f2-4add-aa27-cc7fa78f5da1

1. Cursor still jumps.
2. Resizing window is broken.

logiclrd commented:
Thanks 🙂 It looks like maybe the coordinate systems _are_ mismatched. Also, the suppression of edge scrolling during resizing depends on it getting continuous window resized events during the resize operation. It's one possibility that that isn't happening on OS X, maybe...

logiclrd committed:
Reintroduced the coordinate system translation in MainFlutterWindow.swift.

logiclrd committed:
Updated edgeScrollMouse in model.dart to add a "safe zone" around the window frame that doesn't trigger edge scrolling.

logiclrd committed:
Updated the bumpMouse handler in MainFlutterWindow.swift to call CGAssociateMouseAndMouseCursorPosition to cancel event suppression.

logiclrd committed:
Added debug annotation to the onWindowResized event in tabbar_widget.dart.

logiclrd commented:
Huh :stuck_out_tongue: The documentation for `CGAssociateMouseAndMouseCursorPosition` says to pass `true` in, but apparently that's the wrong data type.

logiclrd committed:
Fix parameter type for CGAssociateMouseAndMouseCursorPosition in MainFlutterWindow.swift.

logiclrd committed:
tabbar_widget.dart: onWindowResized -> onWindowResize

logiclrd committed:
Removed temporary diagnostic debugPrint from tabbar_widget.dart.

logiclrd commented:
Apparently there _is_ a way to get the mouse position within the Core Graphics API. It doesn't seem to be documented, but a) `CGEventGetLocation` returns the coordinates of a mouse event, and b) `CGEventCreate(NULL)`, documented as returning "A new event to be filled in", actually returns an event _that has the current mouse cursor position already filled in_. Done this way, I believe there's no mismatch in the coordinate system.

logiclrd committed:
Updated MainFlutterWindow.swift to obtain the mouse position by creating a dummy CGEvent. The old NSEvent.mouseLocation code is left as a fallback.

logiclrd committed:
The documentation said to be sure to call CFRelease, but apparently it's a build error to do so. :-P

logiclrd committed:
Replaced CGEvent calls in MainFlutterWindow.swift with uses of the CGEvent wrapper struct.

logiclrd committed:
Added argument label to call to CGEvent.init.

logiclrd committed:
Changed mouseLoc from piecewise assignment to assignment of the whole structure, as it is not yet initialized at that point.

fufesou commented:
This PR works perfectly on Windows and macOS now.

> > On Linux, it's too fast for precise adjustments but beneficial for selecting entire lines.
> 
> It sounds like `bumpMouse` isn't working in your configuration. The scrolling speed should be limited to _actual movement of the mouse_, because once the mouse enters the edge scrolling region, it gets warped out of it immediately. What you're describing sounds like the mouse fails to warp out of the edge scrolling region, so each subsequent mouse event repeats the scrolling. In my original video in the PR description, if you watch closely you'll see that each time the mouse bumps into the edge, it gets pushed back -- in essence, the mouse is scrolled _with_ the underlying view.

Yes.
**X11** works fine, **Wayland** has this issue.

Maybe https://github.com/wxWidgets/wxWidgets/pull/25721 helps, but it's a little complicated.

I can also see the print on Linux sometimes.
```
[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: Null check operator used on a null value
#0      RemotePage.ffi (package:flutter_hbb/desktop/pages/remote_page.dart:64:35)
#1      _DesktopTabState.onWindowResize (package:flutter_hbb/desktop/widgets/tabbar_widget.dart:429:46)
#2      DesktopMultiWindow._windowMethodCallHandler (package:desktop_multi_window/desktop_multi_window.dart:129:27)
#3      DesktopMultiWindow.setMethodHandler.<anonymous closure> (package:desktop_multi_window/desktop_multi_window.dart:78:9)
#4      MethodChannel._handleAsMethodCall (package:flutter/src/services/platform_channel.dart:571:55)
#5      MethodChannel.setMethodCallHandler.<anonymous closure> (package:flutter/src/services/platform_channel.dart:564:34)
#6      _DefaultBinaryMessenger.setMessageHandler.<anonymous closure> (package:flutter/src/services/binding.dart:618:35)
#7      _invoke2 (dart:ui/hooks.dart:344:13)
#8      _ChannelCallbackRecord.invoke (dart:ui/channel_buffers.dart:45:5)
#9      _Channel.push (dart:ui/channel_buffers.dart:135:31)
#10     ChannelBuffers.push (dart:ui/channel_buffers.dart:343:17)
#11     PlatformDispatcher._dispatchPlatformMessage (dart:ui/platform_dispatcher.dart:750:22)
#12     _dispatchPlatformMessage (dart:ui/hooks.dart:257:31)
```


There are also too many prints like
```
flutter: [Main] call bump_mouse with args {dx: 55, dy: 0} from window 1
flutter: [Window Main] bump mouse cursor by (55, 0)
flutter: [Main] call bump_mouse with args {dx: 63, dy: 0} from window 1
flutter: [Window Main] bump mouse cursor by (63, 0)
```

Perhaps we can filter these outputs by the event name, as we may not need to print these outputs anymore.

logiclrd commented:
Okay, I've found about about `wayland-scanner` and how libwayland doesn't distribute headers but rather XML definitions of Wayland protocols, and then your build incorporates a step translating the XML definitions into headers and glue code as needed. How does that interact with the automated RustDesk builds? Do they have access to `wayland-scanner`? I guess maybe new associated dependencies need to be added to the build agent initialization?? I saw reference to a Rust version of the scanner, `wayland-rs`, but it doesn't seem to be involved presently in Wayland-specific code in RustDesk.

fufesou commented:
> So it seems to me that:
> 
> 1. We need to try to use `pointer_warp_v1`.
> 2. If we can't, we need to try to use `lock_pointer`.
> 3. If neither of those works, we need to push the failure back up to the UI and forcibly switch the scrolling style away from `scrolledge`, because we don't have the necessary support for it to work properly.

Maybe we only need to use `pointer_warp_v1`.

Because it works on Wayland, but it's too fast for precise adjustments.


> How does that interact with the automated RustDesk builds? Do they have access to `wayland-scanner`?

I have no idea too.

Maybe there is no need to call wayland APIs in rust, you only need to change the `cc` file.

You may need to modify the CI task on Linux. https://github.com/rustdesk/rustdesk/blob/master/.github/workflows/flutter-build.yml 

Run `wayland-scanner` with the XML file, and then compile the project.

But there may be some version errors.
RustDesk still uses `ubuntu 18.04` to run CI https://github.com/rustdesk/rustdesk/blob/e66d2facd461a57ad24d77c50cf1515c4eee025a/.github/workflows/flutter-build.yml#L1289

logiclrd commented:
> Maybe we only need to use `pointer_warp_v1`.
> 
> Because it works on Wayland, but it's too fast for precise adjustments.

Mm, except that it's not working as designed.

The design is that whatever mouse movement you do is captured virtually, so that if the cursor bumps into the edge scrolling region, then the underlying canvas _including the position of the mouse cursor_ is scrolled accordingly. This can only work if Bump Mouse is functional.

I'm working on code that detects that Bump Mouse wasn't able to do its thing, and if so, I'm thinking maybe instead of just disabling edge scrolling entirely, it can fall back on a mode where the scroll velocity is a constant, rather than it being based on how far the mouse has moved into the edge scrolling region.

But there is literally only one compositor right now that supports `pointer_warp_v1`.

Which also means that to test this, it will be necessary to configure a Linux system using Wayland compositor "Jay".

> Maybe there is no need to call wayland APIs in rust, you only need to change the `cc` file.

Yep, that's my plan. This will be embedded in the `linux` platform-specific code underlying the Flutter app. Basically, where the current `bump_mouse` is implemented but only works on X11, that implementation will be broadened to support Wayland.

But, done properly, it's still necessary to integrate `wayland_scanner` into the build process. With most APIs, you just locate their `.h` files and `#include` them, but Wayland _doesn't have `.h` files_. At least, if I'm understanding correctly, it has XML protocol definitions and a code generator, and you're expected to autogenerate transient glue code within the build. Fun! 😁 

> But there may be some version errors.
> RustDesk still uses `ubuntu 18.04` to run CI

More fun 😁 Well, one thing at a time. First step: See it working on my local machine.

logiclrd committed:
Linux platform channel: Refactored bump_mouse, setting the stage for a future Wayland implementation.
- Made a new top-level bump_mouse method in bump_mouse.cc/.h.
- Moved the X11-specific implementation to bump_mouse_x11 in bump_mouse_x11.cc/h.
Reworked the bumpMouse operation to have a boolean return value:
- Updated bumpMouse in platform_channel.dart to return a Future<bool> instead of a Future<void>.
- Windows platform channel: Updated BumpMouse in win32_desktop.cpp to return a bool value. Updated the method call handler "bumpMouse" branch in flutter_window.cpp to propagate the BumpMouse return value back to the originating MethodCall.
- MacOS platform channel: Updated the "bumpMouse" branch in the method call handler in MainFlutterWindow.swift to pass true or false into the 'result()' call.
- Linux platform channel: Updated the bump_mouse top-level method and its underlying implementation bump_mouse_x11 to return bool values. Updated the "bumpMouse" branch of host_channel_call_handler in my_application.cc to propagate the result value back up the method channel.
- Updated the kWindowBumpMouse branch of the method handler registered in desktop_home_page.dart to propagate a return value from RdplatformChannel.bumpMouse.

logiclrd committed:
Reworked the edge scrolling computations in model.dart to use Vector2 from the vector_math package. Updated pubspec.yaml to declare a dependency on vector_math.

logiclrd committed:
Added an alternative edge scrolling mechanism for when "Bump Mouse" functionality is unavailable:
- Added methods setEdgeScrollTimer and cancelEdgeScrollTimer to model.dart, along with a few state fields.
- Updated edgeScrollMouse to latch the (x, y) coordinate of the last edge scroll event, in case it will be autorepeating.
- Updated edgeScrollMouse to check whether the call to the kWindowBumpMouse method of rustDeskWinManager (and thus the underlying bump_mouse method) succeeded, and to switch to timer-based autorepeat if it fails. Made edgeScrollMouse async to allow awaiting the result of the kWindowBumpMouse method call.
- Updated input_model.dart to call cancelEdgeScrollTimer when a new move event is being processed.
- Updated remote_page.dart to call cancelEdgeScrollTimer when the pointer exits the area represented by the view.

logiclrd commented:
The Wayland work-in-progress: https://github.com/logiclrd/rustdesk/pull/1

logiclrd commented:
To explain a little: I had made changes to support X11 and Wayland, and as part of those changes I reworked the Linux platform back-end, refactoring bump_mouse out into multiple files to allow the code to be kept separated. After determining that the Wayland implementation is a no-go today, I split the Git commit apart, keeping the end result but separating the Wayland-specific stuff into an isolated commit. I also split up other unrelated changes into their own commits.

Everything but the Wayland commit is now on this branch, and the https://github.com/logiclrd/rustdesk/pull/1 PR in my fork shows the Wayland changes on top of these changes.

logiclrd commented:
Should I rebase to resolve the `ca.rs` and `ja.rs` merge conflicts?

logiclrd committed:
Fixed scroll percentage math in edgeScrollMouse in model.dart.

fufesou commented:
> Is RustDesk likely to migrate to Gtk 4 at some point in the future?
> Would be a pity to lose that if it could have future value.
> I guess it's ultimately up to Flutter.

Yes. It's up to Flutter.

> Should I rebase to resolve the `ca.rs` and `ja.rs` merge conflicts?

Yes, please.

fufesou commented on flutter/pubspec.yaml:
```diff
@@ -109,6 +109,7 @@ dependencies:
   xterm: 4.0.0
   sqflite: 2.2.0
   google_fonts: ^6.2.1
+  vector_math: ^2.2.0
```
Build error on my machine (flutter 3.24.5, the version of github CI), maybe you should downgrade to `vector_math: 2.1.4`.

```
Note: vector_math is pinned to version 2.1.4 by flutter_localizations from the flutter SDK.
See https://dart.dev/go/sdk-version-pinning for details.


Because flutter_hbb depends on flutter_localizations from sdk which depends on
  vector_math 2.1.4, vector_math 2.1.4 is required.
So, because flutter_hbb depends on vector_math ^2.2.0, version solving failed.
```

logiclrd replied:
Ahh, I am indeed running a newer Flutter SDK and have a handful of local updates to accomodate that.

fufesou commented on flutter/lib/models/model.dart:
```diff
@@ -2052,6 +2121,119 @@ class CanvasModel with ChangeNotifier {
+        setEdgeScrollTimer();
+
+        // When autorepeating, scale down the extent of the scroll, otherwise it'll be way too fast.
+        encroachment *= 0.2;
```
My testing shows that `0.2` is still fast.

`0.05` seems acceptable to me. What do you think?

logiclrd replied:
Mm, interesting. On my end, 0.2 felt exactly right. Ultimately, it will depend on how precise the timer is. The code currently requests a 10ms interval. Maybe on my system, it is in practice getting a longer delay?? Weird that it would be different.

I'm pretty sure 0.05 would be glacially slow on my end... Let me do some testing. Maybe an algorithm that automatically tunes based on feedback from the system high-precision timer is needed.

logiclrd committed:
Fixed declared return value for Win32Desktop::BumpMouse in win32_desktop.h.

logiclrd committed:
Fixed vector_math dependency version in pubspec.yaml to be compatible with the codebase standard Flutter version.

logiclrd committed:
Added class EdgeScrollFallbackState to model.dart for tracking the state of the edge scroll fallback strategy. Factored out the actual edge scrolling action from CanvasModel.edgeScrollMouse to new method performEdgeScroll so that EdgeScrollFallbackState can call it. Updated edgeScrollMouse to not call performEdgeScroll when it's enabling the fallback strategy.
Updated CanvasModel to use EdgeScrollFallbackState instead of directly tracking the state. Removed method setEdgeScrollTimer.
Added method initializeEdgeScrollFallback to CanvasModel that takes a TickerProvider. Updated _RemotePageState to include the mixin TickerProviderStateMixin. Updated _RemotePageState.initState to call canvasModel.initializeEdgeScrollFallback.
Updated handlePointerDevicePos in input_model.dart to not call cancelEdgeScrollTimer before edgeScrollMouse.
Renamed CanvasModel.cancelEdgeScrollTimer to CanvasModel.cancelEdgeScroll.
Updated the calculations in CanvasModel.edgeScrollMouse to only factor in the safe zone if BumpMouse is working. (Otherwise the problem with resizing can't possibly occur.)

logiclrd committed:
Updated CanvasModel.edgeScrollMouse in model.dart to handle the situation where only one of the scrollbars is active. Factored extraction of scrollbar data into new function getScrollInfo.

logiclrd committed:
Updated onWindowResize in tabbar_widget.dart to be resilient to RemotePage instances that don't yet have an ffi reference. Added property hasFFI to remote_page.dart.

logiclrd committed:
Removed debug output from model.dart.

fufesou commented:
> The [debugPrint](https://api.flutter.dev/flutter/rendering/debugPrint.html) function logs to console even in [release mode](https://docs.flutter.dev/testing/build-modes#release). As per convention, calls to [debugPrint](https://api.flutter.dev/flutter/rendering/debugPrint.html) should be within a debug mode check or an assert:
> 

Yes, it's interesting.

> Do we want to keep the debug output in the inter-isolate channel defined in `desktop_home_page.dart` for other calls even in release builds, or would it be appropriate for _all_ of that output to be debug-mode-only?

No need to change the other prints.

The bump cursor logs should be filtered due to their excessive frequency, which leads to confusion.


These prints are not visible to the users unless they run RustDesk via the command line.

logiclrd commented:
This is what I have in mind:

<img width="1198" height="208" alt="image" src="https://github.com/user-attachments/assets/6a63676a-9710-4f40-b6c0-b79a0dd5460a" />

But if you'd rather the `debugPrint` be kept for release builds, just not for `kWindowBumpMouse` calls, then I'll do that. 🙂

logiclrd commented on flutter/lib/models/model.dart:
```diff
@@ -2052,6 +2169,126 @@ class CanvasModel with ChangeNotifier {
+
+    scrollPixel += delta;
+
+    scrollPixel.clamp(Vector2.zero(), max);
```
There is no wrong method call. In the preceding comment, the AI hallucinated a method `clampVector` parallel to `clampScalar`. The `clampScalar` method is real, but the one that takes `Vector2` parameters is just called `clamp`. Note that this line is calling _the same method_ as the one it thinks is wrong on line 2243. 😛

logiclrd committed:
PR feedback:
- Added filtering to diagnostic output in the method handler in desktop_home_page.dart to exclude the very chatty kWindowBumpMouse-related output.
- Removed the diagnostic output from bumpMouse in platform_channel.dart for the same reason.
- Updated setScrollPercent to coalesce NaN values for x and y to 0.
- Initialized the GError pointer variable passed into fl_method_call_respond_success in linux/my_application.cc to NULL.
- Added bounds checking of the argument values in the EncodableList branch of the "bumpMouse" method call handler in windows/runner/flutter_window.cpp.

logiclrd committed:
Added a latch mechanism that keeps edge scrolling disabled until the cursor is observed to be in the inner area bounded by the edge scroll areas:
- Added tristate enumerated type EdgeScrollState to model.dart. In addition to inactive and active states, there is state armed which behaves like inactive but can transition to active when conditions are met.
- Added a field to CanvasModel of type EdgeScrollState. Added methods disableEdgeScroll and rearmEdgeScroll.
- Updated enterView to call canvasModel.rearmEdgeScroll and leaveView to call canvasModel.disableEdgeScroll in remote_page.dart.
- Updated edgeScrollMouse to check the state, disabling edge scrolling when the state is not active and transitioning from armed to active when the mouse is in the interior space.
- Removed the notifyResize/_suppressEdgeScroll mechanism from CanvasModel in model.dart as it is no longer necessary.
- Removed the "safe zone" mechanism from CanvasModel.edgeScrollMouse in model.dart as it is no longer necessary.
- Switched the onWindowResize handler in DesktopTabState in tabbar_widget.dart back to onWindowResized, now that it is no longer delivering canvasModel.notifyResize to all RemotePage tabs.

logiclrd committed:
Fixed memory leak: Added call to free GError object returned by Flutter API in the event of an error.

logiclrd commented on flutter/lib/models/model.dart:
```diff
@@ -2052,6 +2169,131 @@ class CanvasModel with ChangeNotifier {
+    var scrollPixelPercent = scrollPixel.clone();
+
+    scrollPixelPercent.divide(max);
+    scrollPixelPercent.scale(100.0);
```
That's okay. NaN's are coalesced to 0.0 in `setScrollPercent`.

logiclrd commented on flutter/lib/models/model.dart:
```diff
@@ -1789,6 +1832,59 @@ class ViewStyle {
+    }
+  }
+
+  start() {
```
I just looked it up, and if you leave off any return type annotation, the method implicitly returns `dynamic`. With no `return` statements, the returned value is `null`.

Before these changes, `CanvasModel` already has 15 methods implicitly returning `dynamic`. It has two that explicitly specify `void`.

It seems out of scope to fix up the class in this PR, but perhaps it would make sense to have a work item in the backlog to identify all methods that _should_ have `void` return type and fix them.

I'll add `void` annotations to all methods added by this PR, though, for sure.

fufesou commented on flutter/lib/models/model.dart:
```diff
@@ -1810,6 +1906,15 @@ class CanvasModel with ChangeNotifier {
+  // to avoid hammering a non-functional Bump Mouse
+  bool _bumpMouseIsWorking = true;
+  // briefly set to true when the canvas detects a resize
+  bool _suppressEdgeScroll = false;
```
`_suppressEdgeScroll` is never set, is the comment above wrong?

<img width="1212" height="1074" alt="Image" src="https://github.com/user-attachments/assets/5457ce77-2106-484d-9a9c-06c779361079" />


The `Flutter` extension in VS Code can help the Dart diagnostics. https://dart.dev/tools/diagnostics

logiclrd replied:
It's not fundamentally wrong but it's out-of-date, so it's wrong _now_. 🙂

fufesou commented on flutter/lib/models/model.dart:
```diff
@@ -1789,6 +1832,59 @@ class ViewStyle {
+    } else {
+      final thisTickElapsed = totalElapsed - _lastTotalElapsed;
+
+      const double FrameTime = 1000.0 / 60.0;
```
Suggestion from the `Flutter` extension in vscode.

```
The constant name 'FrameTime' isn't a lowerCamelCase identifier.
Try changing the name to follow the lowerCamelCase style.
```


The following names may be better.

```dart
      const double kFrameTime = 1000.0 / 60.0;
      const double kSpeedFactor = 0.1;
```

logiclrd replied:
Ah, yes, I'd seen that elsewhere and it hadn't clicked that that's the convention for constants.

logiclrd commented on flutter/macos/Runner/MainFlutterWindow.swift:
```diff
@@ -99,6 +99,51 @@ class MainFlutterWindow: NSWindow {
+                    let arg = call.arguments as! [String: Any];
+
+                    let dx = (arg["dx"] as? Int) ?? 0;
+                    let dy = (arg["dy"] as? Int) ?? 0;
```
I find it unclear whether the list argument format handling is actually needed. I can't remember now where I found it, but I defniitely found example code that checked for both map and list argument styles. That's why the Linux and Windows versions are written that way. But, the `macos` variant, as highlighted here, has only used the map style the whole time and has, in practice, worked. But, I suppose I'd better update it to match for consistency.

fufesou commented on flutter/lib/models/model.dart:
```diff
@@ -2052,6 +2169,131 @@ class CanvasModel with ChangeNotifier {
+
+    if (x < edgeThickness) {
+      dxOffset = x - edgeThickness;
+    } else if ((x >= size.width - edgeThickness)) {
```
Should it be?

```dart
} else if (x >= (size.width - edgeThickness)) {
```

logiclrd committed:
PR feedback:
- Copilot: Use type annotations.
- Copilot: Condition to stop edge scrolling when fallback strategy is in use and the mouse is moved back to the centre.
- Copilot: Check FLValue type before calling fl_value_get_int.
- Copilot: Support list-style method channel dispatch in "bumpMouse" handler for macos as the linux and windows implementations already do.
- Naming convention for constants.
- Left-over variable from previous strategy: _suppressEdgeScroll.
- Unnecessary extra parentheses in edge scroll area conditions.

fufesou commented on flutter/lib/models/model.dart:
```diff
@@ -1835,14 +1938,24 @@ class CanvasModel with ChangeNotifier {
   double get devicePixelRatio => _devicePixelRatio;
   Size get size => _size;
   ScrollStyle get scrollStyle => _scrollStyle;
+  bool get suppressEdgeScroll => _suppressEdgeScroll;
```
This line should be removed.

logiclrd committed:
Removed property suppressEdgeScroll referencing now-removed field _suppressEdgeScroll in model.dart.
Removed accidental extra blank line in MainFlutterWindow.swift.

logiclrd committed:
Switched CanvasModel.setScrollPercent to use double.isFinite instead of double.isNaN to test for proper numerical values.

logiclrd commented on flutter/lib/models/model.dart:
```diff
@@ -1835,14 +1938,24 @@ class CanvasModel with ChangeNotifier {
-  setScrollPercent(double x, double y) {
-    _scrollX = x;
-    _scrollY = y;
+  void setScrollPercent(double x, double y) {
```
Ah, good catch. I did some poking, and it looks like the `.isFinite` property is the most useful here; it's `false` for both infinity and NaN.

logiclrd replied:
That's not quite the case. The last commit checks `isFinite`, not `isInfinite`. These are not the strict logical inverse of each other. NaN is not infinite, but it is not finite either. `isFinite` returns `false` for both `Infinity` and `NaN`.

```
Value             isFinite  isInfinite  isNaN
----------------  --------  ----------  -----
(numeric)         true      false       false
infinity          false     true        false
negativeInfinity  false     true        false
nan               false     false       true
```

fufesou replied:
Sorry, I misread.

logiclrd commented on flutter/lib/models/model.dart:
```diff
@@ -2052,6 +2166,133 @@ class CanvasModel with ChangeNotifier {
+
+    encroachment.clamp(-scrollPixel, max - scrollPixel);
+
+    if (encroachment.length == 0) {
```
`isZero` seems to be a hallucination, but `length2` works just as well as `length` and does indeed avoid an unnecessary `sqrt`.

logiclrd committed:
PR feedback:
- Copilot: Use Vector2.length2 instead of Vector2.length to avoid an unnecessary sqrt in comparison with zero.
- Copilot: Baleet unnecessary semicolons from Swift code.

logiclrd committed:
PR feedback:
- Copilot: Check argList.count before indexing it

logiclrd commented:
> Another issue with edge scroll on macOS.
> ..
> The cursor does not appear until it moves out of the window.

The video does not appear for me for whatever reason.

> Since you don't have access to macOS, I will try to fix the problem.

Thanks so much. 🙂

logiclrd committed:
Oops with the semicolons again.

fufesou commented:
> > Another issue with edge scroll on macOS.
> > ..
> > The cursor does not appear until it moves out of the window.
> 
> The video does not appear for me for whatever reason.
> 
> > Since you don't have access to macOS, I will try to fix the problem.
> 
> Thanks so much. 🙂


`activeLocalCursor()` should be called in both `moveDesktopMouse()` and `edgeScrollMouse()`.

You can reproduce this issue by enabling "Show remote cursor".
1. Move cursor on the controlled side.
2. Move cursor in local window.

```
diff --git a/flutter/lib/models/model.dart b/flutter/lib/models/model.dart
index 143129736..8d06c2603 100644
--- a/flutter/lib/models/model.dart
+++ b/flutter/lib/models/model.dart
@@ -2151,19 +2151,7 @@ class CanvasModel with ChangeNotifier {
       notifyListeners();
     }

-    // If keyboard is not permitted, do not move cursor when mouse is moving.
-    if (parent.target != null && parent.target!.ffiModel.keyboard) {
-      // Draw cursor if is not desktop.
-      if (!(isDesktop || isWebDesktop)) {
-        parent.target!.cursorModel.moveLocal(x, y);
-      } else {
-        try {
-          RemoteCursorMovedState.find(id).value = false;
-        } catch (e) {
-          //
-        }
-      }
-    }
+    activeLocalCursor(x, y);
   }

   void initializeEdgeScrollFallback(TickerProvider tickerProvider) {
@@ -2195,7 +2183,25 @@ class CanvasModel with ChangeNotifier {
     return (scrollPixel, max);
   }

+  void activeLocalCursor(double x, double y) {
+    // If keyboard is not permitted, do not move cursor when mouse is moving.
+    if (parent.target != null && parent.target!.ffiModel.keyboard) {
+      // Draw cursor if is not desktop.
+    if (parent.target != null && parent.target!.ffiModel.keyboard) {
+      // Draw cursor if is not desktop.
+      if (!(isDesktop || isWebDesktop)) {
+        parent.target!.cursorModel.moveLocal(x, y);
+      } else {
+        try {
+          RemoteCursorMovedState.find(id).value = false;
+        } catch (e) {
+          //
+        }
+      }
+    }
+  }
+
   void edgeScrollMouse(double x, double y) async {
+    activeLocalCursor(x, y);
+
     if ((_edgeScrollState == EdgeScrollState.inactive)
      || (size.width == 0 || size.height == 0)
      || !(_horizontal.hasClients || _vertical.hasClients)) {
```

fufesou committed:
Edge scroll, active local cursor

Signed-off-by: fufesou <linlong1266@gmail.com>

fufesou committed:
Merge branch 'edge-scrolling' of https://github.com/logiclrd/rustdesk into edge-scrolling

logiclrd commented:
Okay I dug deeper and see that it's setting a semaphore that gets picked up by an observer in the widget body that `_RemotePageState` builds.

logiclrd commented on flutter/lib/desktop/pages/desktop_home_page.dart:
```diff
@@ -760,9 +761,19 @@ class _DesktopHomePageState extends State<DesktopHomePage>
+        case kWindowBumpMouse: return true;
+      }
+
+      return false;
```
Deliberately written as a `switch` because the most likely future update is then reduced to a one-line addition.

logiclrd commented on flutter/lib/models/model.dart:
```diff
@@ -2037,21 +2159,155 @@ class CanvasModel with ChangeNotifier {
+    var encroachment = Vector2(dxOffset, dyOffset);
+
+    var (scrollPixel, max) = getScrollInfo();
+
```
??

logiclrd committed:
PR feedback:
- Copilot: Removed unused property hasFFI from remote_page.dart.
- Copilot: Updated updateScrollStyle in model.dart to be resilient to the possibility of bind.sessionGetScrollStyle returning null.

logiclrd committed:
Factored local cursor updates out of CanvasModel.moveDesktopMouse in model.dart, adding new methods activateLocalCursor and updateLocalCursor.
Updated handlePointerDevicePos in input_model.dart to call canvasModel.updateLocalCursor on every mouse event.
Updated initState in remote_page.dart to schedule a call to canvasModel.activateLocalCursor as a first-image callback.

fufesou commented:
Hi, please also address this discussion  https://github.com/rustdesk/rustdesk/pull/13247#discussion_r2476443545

rustdesk commented on flutter/lib/models/model.dart:
```diff
@@ -2037,21 +2159,155 @@ class CanvasModel with ChangeNotifier {
+    }
+
+    // Trigger scrolling when the cursor is close to an edge
+    const double edgeThickness = 100;
```
> Is a range of `100` too wide for edge scrolling?

I am also curious about this. How do you get `100`? do you have any reference design? Is there any other software used as your reference?

logiclrd replied:
It's arbitrary and could absolutely be changed. That said, the mouse isn't captured, so the only thing keeping it inside the window is Bump Mouse. If the mouse cursor manages to cross over the edge scrolling region in its entirety in a single event, then it escapes the window and that's that. Set to 100, it takes a concerted effort to make this happen. If it were set to e.g. 10, edge scrolling would be almost useless because casual use of the mouse would be constantly escaping the window. So a balance must be struck.

logiclrd replied:
One parting thought -- this is definitely logically related but it also inarguably extends the scope of the PR. Perhaps it would make sense to merge it with a fixed value initially, and follow up with a configurable option in a separate PR? This PR's scope has already ballooned a couple of times. 😛

rustdesk replied:
No, you do not need to add anything in config.rs. Please use bind.mainGetLocalOption and bind.mainSetLocalOption. You can find example in desktop_setting_page.dart.

logiclrd replied:
https://github.com/rustdesk/hbb_common/pull/426

fufesou committed:
Remove duplicated condition checks

Signed-off-by: fufesou <linlong1266@gmail.com>

fufesou committed:
Chore

Signed-off-by: fufesou <linlong1266@gmail.com>

logiclrd committed:
Updated the explanation for rounding away from 0 in edgeScrollMouse in model.dart.
